pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md

```diff
@@ -0,0 +1,82 @@
+Telecom companies as they expect a single IP address to bind to. This address needs to be the same and
+DNS records are not accepted either. We are running in AWS as well, so the reader might ask why we are not
+using Elastic IPs and add them to the services? Good idea, but Telecom operators will not whitelist
+an Elastic IP for you as there are no guarantees that it will belong to your Infrastructure forever.
+
+We are a member of RIPE and do have a small subnet block for our own use, so we thought we could make use of 
+that. As we own the block and AWS supports BYOIR (Bring your own IP-range), we created a special subnet with
+some Kubernetes Nodes in it. This was not enough to make this work since we depended on a service always having
+the same IP attached to it, as well as the node running a pod having some very specific Routes set.
+
+With this scenario in mind we set out to find solutions and all solutions we could think of required haggling
+with Kubernetes.
+
+## Extending Kubernetes
+
+There are several ways to extend Kubernetes. All functionality in Kubernetes is build upon very nice and clean
+public APIs, or to say it with other words: There is no private API magic hidden somewhere. So lets look at
+two ways on how to extend kubernetes.
+
+Possible ways to go forward:
+- Adding a scheduler extender
+- Creating an operator/controller
+
+There are more ways to extend kubernetes, but these two ways will be the one we shall look at. Just for completeness
+you can as well also add another scheduler or change kubernetes itself. However these possibilities have some serious
+downsides.
+
+### Adding a scheduler extender
+The kubernetes scheduler checks for certain requirements before it schedules a pod onto a node. Some of these requirements
+are hard requirements, like cpu, memory and number of pods. Other requirements are more soft, like if the pods are allowed to
+be packed together or in which AZ they are gonna run. All of those requirements are collected and points given to each
+node on how good they meet the requirements. The Node that fits best, gets chosen.
+
+All of that are things the scheduler will do for you automatically, however it is also possible to give the cluster a 
+`KubeSchedulerConfiguration` object that will tell the scheduler to also reach out to a service for additional point scoring.
+The SchedulerConfiguration is a JSON file and for further explanations, please have a look at this 
+excellent [blog post](https://developer.ibm.com/articles/creating-a-custom-kube-scheduler/).
+
+In our case, we could have written a service that checks which IP Addresses are assigned to the Nodes and moved the Pods
+onto those nodes. This would have required us to make sure that those Nodes had all needed IP-Addresses all the time.
+That sounded not very enticing when doing cluster upgrades as it would have needed to be at least a semi-manual process.
```

Comment:
I'm left a bit confused as to whether you implemented a scheduler extender or not.

Reply:
Will make that more clear